proxy/mitmproxy.py

```diff
@@ -0,0 +1,122 @@
+import os
+import sys
+from mozprocess import ProcessHandler
+
+dirname = os.path.dirname(__file__)
+
+
+class MITMProxyBase(object):
+
+    def __init__(self, path, proxy_version, record=True):
+        self.path = path
+        self.record = record
+        self.proxy_version = proxy_version
+        self.mitm_home = os.path.join(os.path.expanduser("~"), ".mitmproxy")
+        self.cert = os.path.join(self.mitm_home, "mitmproxy-ca-cert.cer")
+        self.scripts = os.path.join(dirname, "scripts")
+        self.binary_path = os.path.join(dirname, "utils", "mitm%s" % self.proxy_version)
+
+    @property
+    def binary(self):
+        if "linux" in sys.platform:
+            name = "mitmdump-linux"
+        elif "darwin" in sys.platform:
+            name = "mitmdump-osx"
+        elif "win" in sys.platform:
+            name = "mitmdump-win.exe"
+        return os.path.join(self.binary_path, name)
+
+    def start(self):
+        print "Staring MitmProxy"
+        print " ".join(self.command())
+        self.process = ProcessHandler(self.command(),
+                                      logfile=os.path.join(dirname, "mitmproxy.log"))
+        self.process.run()
+        return self.process
+
+    def stop(self):
+        self.process.kill()
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, *args):
+        try:
+            self.process.wait()
+        finally:
+            self.process.kill()
+
+
+class MITMProxy202(MITMProxyBase):
+
+    def __init__(self, path, record=True):
+        MITMProxyBase.__init__(self, path=path, proxy_version="2.0.2", record=record)
+
+    def command(self):
+        if self.record:
+            command = [
+                self.binary,
+                "--wfile",
+                self.path,
+                "--script",
+                " ".join([os.path.join(self.scripts, "inject-deterministic.py")]),
+            ]
+        else:
+            command = [
+                self.binary,
+                "--replay-kill-extra",
+                "--script",
+                " ".join(
+                    [
+                        os.path.join(self.scripts, "alternate-server-replay-2.0.2.py"),
+                        self.path,
+                    ]
+                ),
+            ]
+        return command
+
+
+class MITMProxy404(MITMProxyBase):
+
+    def __init__(self, path, record=True, port_fw = False):
+        self.port_fw = port_fw
+        MITMProxyBase.__init__(self, path=path, proxy_version="4.0.4", record=record)
```

Comment:
Instead of introducing `port_fw` perhaps we could rename `record` to `mode` and have three options `record`, `replay`, and `forward`.

Reply:
Good idea 
Done